request-cert/k8s_certificates.go

```diff
@@ -0,0 +1,104 @@
+// Copyright 2017 The Cockroach Authors.
+//
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+//     http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or
+// implied. See the License for the specific language governing
+// permissions and limitations under the License.
+//
+// Author: Marc berhault (marc@cockroachlabs.com)
+
+package main
+
+import (
+	"flag"
+	"fmt"
+	"time"
+
+	"github.com/pkg/errors"
+
+	types "k8s.io/apimachinery/pkg/apis/meta/v1"
+	"k8s.io/client-go/kubernetes"
+	certificates "k8s.io/client-go/pkg/apis/certificates/v1beta1"
+	_ "k8s.io/client-go/plugin/pkg/client/auth/gcp"
+	"k8s.io/client-go/tools/clientcmd"
+)
+
+var (
+	kubeConfig = flag.String("kubeconfig", "", "config file if running from outside the cluster")
+)
+
+func kubernetesSigner(csrName string, csr []byte, wantServerAuth bool) ([]byte, error) {
+	// Create a config from the config file, or a InCluster config if empty.
+	config, err := clientcmd.BuildConfigFromFlags("", *kubeConfig)
+	if err != nil {
+		return nil, errors.Wrap(err, "error building kubernetes config")
+	}
+
+	// Create the client.
+	client, err := kubernetes.NewForConfig(config)
+	if err != nil {
+		return nil, errors.Wrap(err, "error creating kubernetes client")
+	}
+
+	keyUsages := []certificates.KeyUsage{
+		certificates.UsageDigitalSignature,
+		certificates.UsageKeyEncipherment,
+		certificates.UsageClientAuth,
+	}
+	if wantServerAuth {
+		keyUsages = append(keyUsages, certificates.UsageServerAuth)
+	}
+
+	// Build the certificate signing request.
+	req := &certificates.CertificateSigningRequest{
+		ObjectMeta: types.ObjectMeta{
+			Name: csrName,
+		},
+		Spec: certificates.CertificateSigningRequestSpec{
+			Groups:  []string{"system:authenticated"},
+			Request: csr,
+			Usages:  keyUsages,
+		},
+	}
+
+	fmt.Printf("Sending create request: %s for %s\n", req.Name, *addresses)
+	resp, err := client.Certificates().CertificateSigningRequests().Create(req)
+	if err != nil {
+		return nil, errors.Wrapf(err, "CertificateSigningRequest.Create(%s) failed", req.Name)
+	}
+
+	fmt.Printf("Request sent. To approve, run 'kubectl certificate approve %s'\n", req.Name)
+
+	// TODO(mberhault): we may want a timeout here, perhaps also retries.
```

Comment:
Yeah, I need to figure out which errors are retryable ones. I'll leave that as a todo for now.
Added a 30s log entry to show that we're still waiting.
Here's a sample output with >1m before I ran `kubectl certificate approve ...`:
```
$ kubectl logs cockroachdb-0 -c init-certs
+ hostname -f
+ /request-cert -certs-dir=/cockroach-certs -type=node -addresses=localhost,127.0.0.1,172.17.0.4,cockroachdb-0.cockroachdb.default.svc.cluster.local,cockroachdb-public -symlink-ca-from=/var/run/secrets/kubernetes.io/serviceaccount/ca.crt
W0622 13:50:40.518620       6 client_config.go:525] Neither --kubeconfig nor --master was specified.  Using the inClusterConfig.  This might not work.
Sending create request: node-cockroachdb-0 for localhost,127.0.0.1,172.17.0.4,cockroachdb-0.cockroachdb.default.svc.cluster.local,cockroachdb-public
Request sent. To approve, run 'kubectl certificate approve node-cockroachdb-0'
2017-06-22 13:51:10.580442775 +0000 UTC: waiting for 'kubectl certificate approve node-cockroachdb-0'
2017-06-22 13:51:40.597192333 +0000 UTC: waiting for 'kubectl certificate approve node-cockroachdb-0'
request node-cockroachdb-0 Approved at 2017-06-22 13:51:47 +0000 UTC
  reason:   KubectlApprove
  message:  This CSR was approved by kubectl certificate approve.
wrote key file: /cockroach-certs/node.key
wrote certificate file: /cockroach-certs/node.crt
symlinked CA certificate file: /cockroach-certs/ca.crt -> /var/run/secrets/kubernetes.io/serviceaccount/ca.crt
```